nomad/deploymentwatcher/deployments_watcher.go

```diff
@@ -191,6 +190,21 @@ func (w *Watcher) watchDeployments(ctx context.Context) {
 				w.remove(d)
 			}
 		}
+
+		// Ensure we're not tracking deployments that have been deleted because
```

Comment:
If you check for the number of watchers and number of deployments, you could avoid always checking even when there are no changes

Reply:
Good catch. That also made me realize that we're not locking the read on `w.watchers` here, so I need to fix that.

Reply:
I don't think we can't optimize this with a length check after all. Imagine the following scenario:

* watchers for deployment A and B exist, len(deployments) == 2, len(watchers) == 2
* A gets purged, B completes (is no longer active), C gets added
* after we call `add`/`remove`, we have len(deployments) == 2, len(watchers) == 2
* so now checking the lengths would be equal and we'd skip removing A

I've pulled this block of code out to a function to make the locking nicer, and I've added a note about why we can't optimize it with the length check.